lib/github-pages-health-check/domain.rb

```diff
@@ -0,0 +1,202 @@
+module GitHubPages
+  module HealthCheck
+    class Domain < GitHubPages::HealthCheck::Checkable
+
+      attr_reader :host
+
+      def initialize(host)
+        @host = host_from_uri(host)
+      end
+
+      # Runs all checks, raises an error if invalid
+      def check!
+        raise Errors::InvalidDNSError unless dns_resolves?
+        return true if proxied?
+        raise Errors::DeprecatedIPError if deprecated_ip?
+        raise Errors::InvalidARecord    if invalid_a_record?
+        raise Errors::InvalidCNAME      if invalid_cname?
+        raise Errors::NotServedByPages  unless served_by_pages?
+        true
+      end
+
+      def deprecated_ip?
+        return @deprecated_ip if defined? @deprecated_ip
+        @deprecated_ip = a_record? && old_ip_address?
+      end
+
+      def invalid_a_record?
+        return @invalid_a_record if defined? @invalid_a_record
+        @invalid_a_record = valid_domain? && a_record? && !should_be_a_record?
+      end
+
+      def invalid_cname?
+        return @invalid_cname if defined? @invalid_cname
+        @invalid_cname = begin
+          return false unless valid_domain?
+          !github_domain? && !apex_domain? && !pointed_to_github_user_domain?
+        end
```

Comment:
This could be simplified to

```ruby
@invalid_cname = !valid_domain? || !(github_domain? || apex_domain? || pointed_to_github_user_domain?)
```

I flipped the `!... && !..` to `!(... || ...)` for slightly easier reading. Feel free to use `(!.... && !....)` if it's easier to read the other way.